esp/services/ws_workunits/ws_wudetails.cpp

```diff
@@ -0,0 +1,349 @@
+{
+    bool isNew;
+    ForEachItemIn(idx,returnAttriblist)
+    {
+        const char * attribName = returnAttriblist[idx];
+        if (*attribName==0) continue;
+
+        StatisticKind sk = queryStatisticKind(attribName);
+        if (sk==StKindNone || sk==StKindAll)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid Attribute name in AttributeToReturn(%s)",attribName);
+
+        unsigned short usk = static_cast<unsigned short> (sk);
+        skReturnAttribList.bAdd(usk, funcCompareUnsignedShort, isNew);
+    }
+
+}
+
+void CWUDetails::buildScopeTypeList(UnsignedShortArray & processedScopeTypeList,
+                                    StringArray & arrayScopeTypeList)
+{
+    bool isNew;
+    ForEachItemIn(idx, arrayScopeTypeList)
+    {
+        const char * scopeType = arrayScopeTypeList[idx];
+        if (*scopeType==0) continue;
+
+        StatisticScopeType sst = queryScopeType(scopeType);
+        if (sst==SSTnone ||sst==SSTall)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid ScopeType(%s) in filterNestedScopeTypes",scopeType);
+
+        unsigned short usst = static_cast<unsigned short> (sst);
+        processedScopeTypeList.bAdd(usst, funcCompareUnsignedShort, isNew);
+    }
+}
+
+void CWUDetails::processWUStatsIntoResponse(IArrayOf<IEspWUScope> & respScopes,
+                                            IConstWUStatisticIterator & statsIter,
+                                            UnsignedShortArray & nestedScopeTypes,
+                                            UnsignedShortArray & returnAttribList,
+                                            StringArray & ids,
+                                            IArrayOf<CAttributeFilter> & arrayAttributeFilter,
+                                            unsigned __int64 minVersion,
+                                            unsigned __int64 & maxTimestamp)
+{
+    unsigned maxDepth = nfilter.getDepth();
+    Owned<CAttributeFilter> tmpForAttribSearch = new CAttributeFilter;
+    bool limitToIdList = ids.ordinality()>0;
+    bool returnSpecifyReturnAttribs = returnAttribList.ordinality()>0;
+    bool filterNestedScopeTypes = nestedScopeTypes.ordinality()>0;
+    bool filterByAttributeValue = arrayAttributeFilter.ordinality();
+
+    DBGLOG("Id list count: %d", ids.ordinality());
+    DBGLOG("ReturnAttribute  list count: %d", returnAttribList.ordinality());
+    DBGLOG("filterNestedScopeTypes count: %d", nestedScopeTypes.ordinality());
+    DBGLOG("AttributeFilter count: %d", arrayAttributeFilter.ordinality());
+
+    ids.sortAscii();      // Will need to lookup up 'id' in ids (for bSearch)
+    IArrayOf<IEspWUResponseAttribute> attribs;
+    SCMStringBuffer currentScope,currentScopeType,currentId;
+    unsigned statsCount = 0, statsAdded = 0;
+    ForEach(statsIter)
+    {
+        ++statsCount;
+        IConstWUStatistic & wuStat = statsIter.query();
+
+        SCMStringBuffer wuScope,  wuId;
```

Comment:
note sure why whese have wu prefixes.  It implies they are something to do with the workunit, which isn't true.  The code is clearer without them.